intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp

```diff
@@ -2052,6 +2052,78 @@ void jblas_reordered_attn_fp32_shift_rope_k(char* cache, const ne_fp16_t* cossin
     }
 }
 
+template <bool zero_padding>
+void jblas_fusion_attn_fp32_batch_cpy_k_(const jblas_fusion_attn_fp32_batch_cpy_kv_args_t* params) {
+  const auto p = *params;
+  const auto pad_headsize = padto(p.head_size, 32);
+  const auto pad_seq_max = padto(p.seq_max, 48);
+  const auto step_head_num = pad_headsize * pad_seq_max;
+
+  const auto size_unaligned_cpy = std::min(padto(p.seq_off, 48) - p.seq_off, p.seq_size);
+  const auto size_aligned_cpy = pad_headsize * (padto(p.seq_off + p.seq_size, 48) - padto(p.seq_off, 48));
+#pragma omp parallel for
+  for (int ihn = 0; ihn < p.heads_kv; ++ihn) {
+    const auto dst = reinterpret_cast<bf16*>(p.dst) + ihn * step_head_num;
+    const auto src = reinterpret_cast<bf16*>(p.src) + ihn * step_head_num;
+
+    if (size_unaligned_cpy) {
+      for (int j = 0; j < pad_headsize; j += 2) {  // K-dim padding for QK_GEMM
+        const auto ii = p.seq_off % 48;
+        const auto i_blk = p.seq_off - ii;
+        const auto off = i_blk * pad_headsize + ii * 2 + j * 48;
+        memcpy(dst + off, src + off, sizeof(bf16) * 2 * size_unaligned_cpy);
+      }
+    }
+    if constexpr (zero_padding) {
+      if (size_aligned_cpy) {
+        const auto off = padto(p.seq_off, 48) * pad_headsize;
+        memcpy(dst + off, src + off, sizeof(bf16) * size_aligned_cpy);
+      }
+    } else {
+      assert(("Unimplemented!", false));
+    }
+  }
+}
+void jblas_fusion_attn_fp32_batch_cpy_k(const jblas_fusion_attn_fp32_batch_cpy_kv_args_t* params) {
+  return params->no_zeroing ? jblas_fusion_attn_fp32_batch_cpy_k_<false>(params)
+                            : jblas_fusion_attn_fp32_batch_cpy_k_<true>(params);
+}
+
+template <bool zero_padding>
+void jblas_fusion_attn_fp32_batch_cpy_v_(const jblas_fusion_attn_fp32_batch_cpy_kv_args_t* params) {
+  const auto p = *params;
+  const auto pad_headsize = padto(p.head_size, 48);
+  const auto pad_seq_max = padto(p.seq_max, 32);
+  const auto step_head_num = pad_headsize * pad_seq_max;
+
+  const auto seq_off_aligned = padto(p.seq_off, 2);
+  const auto seq_end_aligned = padto(p.seq_off + p.seq_size, 32);
+  const auto seq_size_aligned = seq_end_aligned - seq_off_aligned;
+#pragma omp parallel for collapse(2)
+  for (int ihn = 0; ihn < p.heads_kv; ++ihn) {
+    for (int j = 0; j < p.head_size; j += 48) {
+      const auto dst = reinterpret_cast<bf16*>(p.dst) + ihn * step_head_num + pad_seq_max * j;
+      const auto src = reinterpret_cast<bf16*>(p.src) + ihn * step_head_num + pad_seq_max * j;
+      if (p.seq_off != seq_off_aligned) {
+        const auto off = (seq_off_aligned - 2) * 48 + 1;
+        for (int jj = 0; jj < 48; ++jj) src[off + jj * 2] = dst[off + jj * 2];
```

Comment:
I think this will greatly hurt performance, any better solution?

Reply:
maybe mask load is better since 48 is static

Reply:
> I think this will greatly hurt performance, any better solution?

Talking about performance, we need a way to profile non-eval cost.

Reply:
I think the cost is neglectable and it don't worth the effort to optimized it. There are only (head_size elements) process in this way vs (seq_size*head_size elements) processed in `memcpy`.

Reply:
But it seems to be a bug here! I should copy src to dst

Reply:
> > maybe mask load is better since 48 is static
> 
> I think the cost is neglectable and it don't worth the effort to optimized it. There are only (head_size elements, note the outer loop is in step of 48) process in this way vs (seq_size*head_size elements) processed in `memcpy`.

@yuchengliu1 can you provide a unified vectorized function to simplify these effort?

Reply:
> But it seems to be a bug here! I should copy src to dst

then why the output are correct? or they not?

Reply:
Maybe only errors of one token don't affect that much

Reply:
For anyone who ran into here, you can manipulate compiler carefully https://godbolt.org/z/o6EEz9x44

Reply:
> But it seems to be a bug here! I should copy src to dst

Fixed in 92ff74ad17680cee6a0977f1626d86eea7bfad03